File: lib/data/mappers/user_mapper.dart
import '../database/app_database.dart';
import 'package:drift/drift.dart';
import '../../domain/entities/user_entity.dart';
import '../../domain/enums/user_status.dart';

extension UserMapper on UserRow {
  UserEntity toDomain() {
    return UserEntity(
      id: id,
      email: email,
      name: name,
      role: role,
      status: UserStatus.values.firstWhere(
        (s) => s.name == status,
        orElse: () => UserStatus.inactive,
      ),
      approvedAt: approvedAt,
      approvedBy: approvedBy,
      lastLoginAt: lastLoginAt,
      avatarUrl: avatarUrl,
      // Sync mappings
      createdAt:
          createdAt, // Table has withDefault(currentDateAndTime) so it should be non-null in Row
      updatedAt: updatedAt ?? createdAt,
      firebaseId: firestoreUid ?? remoteId,
      createdBy: null,
      modifiedBy: null,
    );
  }
}

extension UserEntityCompanionMapper on UserEntity {
  UsersCompanion toCompanion() {
    return UsersCompanion(
      email: Value(email),
      firestoreUid: Value(firebaseId),
      name: Value(name),
      role: Value(role),
      status: Value(status.name),
      approvedAt: Value(approvedAt),
      approvedBy: Value(approvedBy),
      lastLoginAt: Value(lastLoginAt),
      avatarUrl: Value(avatarUrl),
      createdAt: Value(createdAt),
      updatedAt: Value(updatedAt),
    );
  }
}

File: courtcare-web/src/data/mappers/user.mapper.ts
import { User } from '../../domain/entities/user';
import { Role, UserStatus } from '../../domain/enums';
import { Timestamp } from 'firebase/firestore';

function parseTimestamp(ts: unknown): Date {
  if (ts instanceof Timestamp) {
    return ts.toDate();
  }
  if (typeof ts === 'string') {
    return new Date(ts);
  }
  return new Date();
}

export function firestoreToUser(id: string, data: Record<string, unknown>): User {
  return {
    id: Number(data['id'] ?? 0),
    email: String(data['email'] ?? ''),
    name: String(data['name'] ?? ''),
    role: Object.values(Role).includes(data['role'] as Role)
      ? (data['role'] as Role)
      : Role.AGENT,
    status: Object.values(UserStatus).includes(data['status'] as UserStatus)
      ? (data['status'] as UserStatus)
      : UserStatus.INACTIVE,
    lastLoginAt: data['lastLoginAt'] != null ? parseTimestamp(data['lastLoginAt']) : null,
    avatarUrl: data['avatarUrl'] != null ? String(data['avatarUrl']) : null,
    approvedAt: data['approvedAt'] != null ? parseTimestamp(data['approvedAt']) : null,
    approvedBy: data['approvedBy'] != null ? String(data['approvedBy']) : null,
    createdAt: parseTimestamp(data['createdAt']),
    updatedAt: parseTimestamp(data['updatedAt']),
    firebaseId: id,
    createdBy: data['createdBy'] != null ? String(data['createdBy']) : null,
    modifiedBy: data['modifiedBy'] != null ? String(data['modifiedBy']) : null,
  };
}

export function userToFirestore(user: User): Record<string, unknown> {
  const result: Record<string, unknown> = {
    email: user.email,
    name: user.name,
    role: user.role,
    status: user.status,
    createdAt: Timestamp.fromDate(user.createdAt),
    updatedAt: Timestamp.fromDate(user.updatedAt),
    firebaseId: user.firebaseId,
  };

  if (user.lastLoginAt !== null) result['lastLoginAt'] = Timestamp.fromDate(user.lastLoginAt);
  if (user.avatarUrl !== null) result['avatarUrl'] = user.avatarUrl;
  if (user.approvedAt !== null) result['approvedAt'] = Timestamp.fromDate(user.approvedAt);
  if (user.approvedBy !== null) result['approvedBy'] = user.approvedBy;
  if (user.createdBy !== null) result['createdBy'] = user.createdBy;
  if (user.modifiedBy !== null) result['modifiedBy'] = user.modifiedBy;

  return result;
}
